### ISSUE TITLE
Memory bug? VectorAdd example fails

### ISSUE THREAD

**Issue by tdd11235813** [ComputationalRadiationPhysics/alpaka#504]
When using CUDA for vectorAdd example, memory glitches occur for non-small numElements.
Result is not computed for every 3rd row (as Elems = 3). For `numElements=123456` the first ~70k-100k rows are correct, but then the results are missing in every `Elems` row. The start index is random.

- when using printf inside the kernel or when using nvprof or cuda-memcheck, this does not happen
- CUDA volatile pointers or synchronizing the global writes by `__syncthreads` does not help
- the C buffer is initialized with 0s at the beginning

Possible output:
```
C[102624] == 0 != 2065228295
C[102627] == 0 != 1325975647
C[102630] == 0 != 2984720098
C[102633] == 0 != 3625720250
C[102636] == 0 != 3210258622
C[102639] == 0 != 4187900660
C[102642] == 0 != 3353799260
C[102645] == 0 != 1166409648
C[102648] == 0 != 1198068419
C[102651] == 0 != 2348462223
// ...
```
Have to dig further into this, it looks quite strange at the moment. [My current hip branch](https://github.com/tdd11235813/alpaka/tree/alpaka-hip).
Tested on Taurus/K80 with CUDA 8.0.61 and CUDA 9.1.85, gcc 5.3, boost 1.64.0. K80 GPUs had no ECC errors.

**Comment by BenjaminW3** [comment]
Each third element does not sound like a memory error. If this is really the number of elements per thread, this might be a strong hint. Maybe there is some computation error in there or maybe some operation overflows when computing the first element for a given thread.

**Comment by tdd11235813** [comment]
ok, it has nothing to do with element or index computation. It is probably caused by the different streams for memory and compute. When there is a `cudaDeviceSynchronize()` before `enqueue`ing the kernel and after the copies, it works. If such a sync is missing, then the kernel in the compute stream starts while memory is still copying.
What buffles me is, the cudaMemcpy is a device-wide sync and the host waits for its completion.

If you want to reproduce, take the current develop branch and change to CUDA backend in the vectorAdd example.
```C++
// in vectorAdd/src/main.cpp
// Define the accelerator                          
//using Acc = alpaka::acc::AccCpuSerial<Dim, Idx>; 
using Acc = alpaka::acc::AccGpuCudaRt<Dim, Idx>;   
using DevAcc = alpaka::dev::Dev<Acc>;              
using PltfAcc = alpaka::pltf::Pltf<DevAcc>;        
//using QueueAcc = alpaka::queue::QueueCpuSync;    
using QueueAcc = alpaka::queue::QueueCudaRtSync;   
// ...
cudaDeviceSynchronize(); // must be called before enqueue, but after the H2D copies 
                         // ... cudaStreamSynchronize(0 /*or queue */) is not enough
// Enqueue the kernel executor
alpaka::queue::enqueue(queue, exec);
```
It also works, if `QueueCudaRtSync` uses the default stream instead of non-blocking.

**Comment by tdd11235813** [comment]
It seems to be a bug in CUDA 8. This issue does not occur on CUDA 9.0 and CUDA 9.1.
```
Currently Loaded Modules:
1) modenv/classic (S)   2) cmake/3.10.1   3) git/2.15.1   4) gcc/4.9.3   5) cuda/9.0.176
#+ own built of Boost 1.65.1 with gcc 4.9.3
```
I also would like to test with clang, but clang<4 is not supported and clang 4 requires CUDA 9 at least to work together.
If this bug is real, then alpaka + CUDA 8 is buggy too. A workaround would be to enforce device-wide synchronization between the stream callbacks.

**Comment by BenjaminW3** [comment]
Do you have any updates on this issue?

**Comment by tdd11235813** [comment]
I gonna recheck it, if the problem persists due to the updates in alpaka.
I started to get a minimal working example, where only the involved alpaka counterparts are implemented to play around with the issue, but did not get it working. With CUDA 9 the issue has been gone and for CUDA 8 you had to synchronize as mentioned above.

**Comment by psychocoderHPC** [comment]
solved with #877
